nodebuilder/node/uptime.go

```diff
@@ -0,0 +1,139 @@
+package node
+
+import (
+	"context"
+	"encoding/binary"
+	"math"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+)
+
+// UptimeMetrics is a struct that records
+//
+//  1. node start time: the timestamp when the node was started
+//  2. node uptime: total time the node has been running counted in units of 1 second
+//
+// the node start time is recorded imperatively when RecordNodeStartTime is called
+// whereas the node uptime is recorded periodically
+// upon callback recalling (re-mettering from optl)
+type UptimeMetrics struct {
+	// nodeStartTS is the timestamp when the node was started.
+	nodeStartTS asyncfloat64.Gauge
+
+	// totalNodeUptime is the total time the node has been running.
+	totalNodeUptime asyncfloat64.Counter
+
+	// totalNodeUpTimeTicks is the total time the node has been running
+	// counted in units of 1 second.
+	totalNodeUpTimeTicks int
+
+	// store is the datastore used to store the node uptime metrics.
+	store datastore.Batching
+}
+
+var (
+	meter          = global.MeterProvider().Meter("node")
+	storePrefix    = datastore.NewKey("node")
+	nodeStartTSKey = datastore.NewKey("node_start_ts")
+)
+
+// NewUptimeMetrics creates a new UptimeMetrics
+// and registers a callback to re-meter the totalNodeUptime metric.
+func NewUptimeMetrics(ds datastore.Batching) (*UptimeMetrics, error) {
+	nodeStartTS, err := meter.
+		AsyncFloat64().
+		Gauge(
+			"node_start_ts",
+			instrument.WithDescription("timestamp when the node was started"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	totalNodeUptime, err := meter.
+		AsyncFloat64().
+		Counter(
+			"node_uptime",
+			instrument.WithDescription("total time the node has been running"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &UptimeMetrics{
+		nodeStartTS:          nodeStartTS,
+		totalNodeUptime:      totalNodeUptime,
+		store:                namespace.Wrap(ds, storePrefix),
+		totalNodeUpTimeTicks: 1,
+	}
+
+	err = meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			totalNodeUptime,
+		},
+		func(ctx context.Context) {
+			m.totalNodeUpTimeTicks++
+			totalNodeUptime.Observe(ctx, float64(m.totalNodeUpTimeTicks))
+		},
+	)
```

Comment:
Looks like you are observing amount of callback calls here, but not the amount of seconds from node start.

Reply:
Correct, noticed this while visually testing this. Fixed it, thank you.